Principal with Key Management

fredfortier opened the pull request:
Notes:
- Deploying the Enigma contract does not work. You'll notice that related sections of the code are commented out. I did not completely remove them because we may want to re-enable them or adapt to your integration test framework. There are two reasons why I chose to disable this: 
  1. Rust Web3 does not support deploying contracts with linked libraries: https://github.com/tomusdrw/rust-web3/issues/186. We could submit a PR to enable it but I determined that it was not a priority.
  2. New business logic like the worker selection algorithm requires extensive setup (registering multiple workers, crossing epoch, etc).

  Here is what I did to compensate:
  1. Added lower level unit tests with mock data that does not interface with the live contract.
  2. Made the CLI commands more granular (which useful beyond testing) and added a mode to `Enigma.specs.js` that invokes the live Principal node. This way, we can substitute the mock for live Principal node and without changing the tests. This was the most convenient way for me to test and also a good bridge to the full integration test script.
- Added a `confirmations` parameter to the `enigma_contract` module in addition to supporting the new methods in the contract.
- I'm using a specific terminology for some of the for the `epoch` and `keys` modules. I'm also using the `_u` and `_t` but there was still overlap: 1) Provider: Provide external data; 2) Keeper: Store and verify data in the enclave. I'm happy to refactor this if you suggest a different convention.
- The `keys_keeper_types` module is copied across: `app/src/esgx/keys_keeper_u.rs` and `enclave/src/keys_keeper_t/keeper_types_t.rs`. I tried to make a reusable module that works both in the enclave and out ran into issues and aborted for now.
- I created `enigma-tools-t/src/document_storage_t.rs` which is based upon `storage_t.rs` but supports variable byte array size. It also has improved error handling. I believe that it can replace `storage_t.rs`.

TODOs:
- Seal the epoch data. I have a plan for this to did not have time to finish. Instead of sealing the entire epoch data, I want to seal the nonce + a hash. The `epoch_provider` will fetch the epoch data and call `set_worker_params` and pass the nonce. Instead of incrementing the nonce, the `epoch_keeper` will verify the data against the hash and store in memory.
- I could not get the Makefile to copy `enigma-types.h` in the enclave folder. I included a temporary workaround on line 111 of `enigma-principal/Makefile`. I would appreciate if you could take a look. This causes the build to fail on the CI server.
- Worker selection is disabled for the time being (it always return the keys). I'm working on testing more thoroughly and will push an update shortly.
- I cherry picked commit 346a8e40827ee4f9af0f56186d72ac916f26cf6f because I needed a working SPID. Now the `attestation_service::service::test` tests are failing. The root cause is not obvious to me. I will wait until the related PR is finalized to troubleshoot.
- Logging is not implemented in the enclave.

Commit:
fixing bug with missing platform info in the report

fredfortier committed:
fixed issue with sgx report

fredfortier committed:
BLD: working on JSON RPC server

fredfortier committed:
BLD: reading the worker param event

fredfortier committed:
BLD: parsing the event log correctly

fredfortier committed:
BLD: experimenting with ecalls

fredfortier committed:
BLD: combined commit of Principal node change for Discovery

fredfortier committed:
BLD: fixed makefile

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	eng_wasm/src/lib.rs
#	enigma-core/app/Cargo.lock
#	enigma-core/app/src/km_u.rs
#	enigma-core/enclave/src/km_t/mod.rs
#	enigma-core/enclave/src/lib.rs
#	enigma-principal/app/Cargo.lock
#	enigma-principal/enclave/src/lib.rs
#	enigma-tools-t/src/common/errors_t.rs
#	enigma-tools-t/src/cryptography_t/asymmetric.rs
#	enigma-types/src/types.rs
#	examples/eng_wasm_contracts/contract_with_eth_calls/Cargo.lock
#	examples/eng_wasm_contracts/contract_with_eth_calls/Cargo.toml
#	examples/eng_wasm_contracts/contract_with_eth_calls/src/lib.rs
#	examples/eng_wasm_contracts/simplest/Cargo.lock
#	examples/eng_wasm_contracts/simplest/Cargo.toml
#	examples/eng_wasm_contracts/simplest/src/lib.rs

fredfortier committed:
BLD: Refactoring for cleaner organization

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-principal/app/Cargo.lock
#	enigma-principal/app/Cargo.toml
#	enigma-principal/app/src/boot_network/deploy_scripts.rs
#	enigma-principal/app/src/boot_network/mod.rs
#	enigma-principal/app/src/boot_network/principal_manager.rs
#	enigma-principal/app/src/boot_network/principal_utils.rs
#	enigma-principal/app/src/cli/options.rs
#	enigma-principal/app/src/esgx/mod.rs
#	enigma-principal/app/src/esgx/random_u.rs
#	enigma-principal/app/src/main.rs
#	enigma-principal/app/tests/principal_node/config/principal_test_config.json
#	enigma-tools-t/Cargo.toml
#	enigma-tools-u/Cargo.toml
#	enigma-tools-u/src/attestation_service/service.rs

fredfortier committed:
BLD: Cleanup after merge

fredfortier committed:
BLD: Post merge adjustments and fixes

fredfortier committed:
DOC: JSON RPC endpoint documentation

fredfortier committed:
DEV: fixed config files and some bugs when running in mining mode

fredfortier committed:
DEV: Made the filter_worker_params fn work again

fredfortier committed:
DEV: Fixed an issue where the random seed generation fired for every block

fredfortier committed:
DEV: WIP - Integrating the backend with the block data

fredfortier committed:
BLD: Encapsulating the epoch stuff so it's not mixed with actual functions of the contract

fredfortier committed:
DEV: Refactored the event filter to support batching

fredfortier committed:
DEV: Experimenting with receipts and blocks validation

fredfortier committed:
DEV: experimenting with block hashing

fredfortier committed:
DEV: Got the JSON-RPC server working at a basic level

fredfortier committed:
DEV: Polishing the JSON-RPC server

fredfortier committed:
DEV: Optimized imports throughout

fredfortier committed:
DEV: Refactored to use RLP to serialize Receipts and Block Headers with enclave. Updagrading to the discovery contract.

fredfortier committed:
DEV: Fixes and refactoring to improve the structure of the modules

fredfortier committed:
DEV: Refactored the event filter to support batching

fredfortier committed:
DEV: Fixes and refactoring to improve the structure of the modules

fredfortier committed:
DEV: Fixes and refactoring to improve the structure of the modules

fredfortier committed:
DEV: Fixes and refactoring to improve the structure of the modules

fredfortier committed:
DEV: Encoding/decoding fixes and improved unit tests

fredfortier committed:
DEV: Receipt serialization fixes and working on verification

fredfortier committed:
Fixed Receipt and Block types returned to the web3 master after merged PR

fredfortier committed:
Testing end-to-end with the new contract, fixes and idiomatic improvements. WIP

fredfortier committed:
Improved ecall error handling and other fixes

fredfortier committed:
Fixed some issues with setting worker parameters and verifying the nonce, unit tests now passing

fredfortier committed:
Fixes and refactoring in set_worker_parameters

fredfortier committed:
DEV: Added a signature when calling `set_worker_parameters` and improved documentation

fredfortier committed:
DEV: Fixed the `get_state_keys` unit tests, some refactoring in the area and disabled selected worker verification

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-principal/app/Cargo.lock
#	enigma-principal/app/Cargo.toml
#	enigma-runtime-t/Cargo.lock
#	enigma-tools-t/src/lib.rs
#	enigma-tools-u/Cargo.toml

fredfortier committed:
DEV: Fixed merge issues with develop and optimized imports

fredfortier committed:
DEV: Fixed an issue with the `get_state_keys` response

fredfortier committed:
DEV: Fixed an issue with the `get_state_keys` response

fredfortier committed:
DEV: Improved my docker startup script

fredfortier committed:
DEV: Re-adjusted the principal config files to point to local contracts and upgraded them

fredfortier committed:
DEV: Minor cleanup edits

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-principal/app/Cargo.lock
#	enigma-runtime-t/Cargo.lock
#	enigma-tools-t/src/cryptography_t/asymmetric.rs
#	enigma-tools-t/src/lib.rs
#	enigma-tools-u/src/web3_utils/enigma_contract.rs
#	enigma-tools-u/src/web3_utils/w3utils.rs
#	enigma-types/src/types.rs

fredfortier committed:
BLD: Removing Cargo.lock files

fredfortier committed:
DEV: Implemented worker selection algo with related unit test

fredfortier committed:
DEV: Created a copy of the storage_t module for sealing. Implemented generics for reusability across data types and improved error handling.

fredfortier committed:
DEV: Error handling cleanup

fredfortier committed:
DEV: minor naming cleanup

fredfortier committed:
DEV: Implemented sealing for `keys_keeper_t`. Passes compiler but more testing required.

fredfortier committed:
DEV: Minor fixes to the state keys storage.

fredfortier committed:
DEV: Added unit test for keys storage

fredfortier committed:
DEV: Adjusted the `getStateKeys` endpoint to include the public encryption key in the response

fredfortier committed:
WIP: Trying to deploy contracts with linked libraries

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-crypto/src/asymmetric.rs
#	enigma-principal/app/Cargo.lock
#	enigma-principal/app/src/main.rs
#	enigma-principal/enclave/Cargo.toml
#	enigma-principal/enclave/src/lib.rs
#	enigma-tools-t/src/common/errors_t.rs
#	enigma-tools-t/src/lib.rs
#	enigma-tools-u/Cargo.toml
#	enigma-tools-u/src/lib.rs
#	enigma-tools-u/src/web3_utils/enigma_contract.rs
#	enigma-tools-u/src/web3_utils/w3utils.rs

fredfortier committed:
Updated the description of the api

fredfortier committed:
Fixed formatting

fredfortier committed:
DEV: Merging with the develop branch

fredfortier committed:
DOC: updated the README file

fredfortier committed:
DEV: trying to fix some build issues. May not build yet, please wait for the next commit.

fredfortier committed:
DEV: Mergin develop

fredfortier committed:
BLD: adjusted the Makefile to include enigma-types

fredfortier committed:
FIX: fixed an issue with the signature recovery unit test

fredfortier committed:
BLD: Quick workaround to include the enigma-types in the enclave module of the principal node. Need to understand how enigma-core does this and replace with the correct logic.

fredfortier committed:
FIX: fixed an issue with the signature recovery unit test

fredfortier committed:
FIX: Finalized develop merge

fredfortier committed:
BLD: Build passing after removing receipt verification

fredfortier committed:
BLD: Unit testing epoch storage

fredfortier committed:
DEV: fixed the register fn

fredfortier committed:
FIX: Fixed issue with reading the Ganache config

fredfortier committed:
FIX: fixed issue with account address config

fredfortier committed:
FIX: Listening to 0.0.0.0 in order play nice with external networks

fredfortier committed:
DEV: Printing the report content for troubleshooting

fredfortier committed:
DEV: Getting the list of active workers from the Enigma contract

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-principal/enclave/src/lib.rs

fredfortier committed:
FIX: Adjustments after merging from develop

fredfortier committed:
FIX: Removed string wrapping in the JSON-RPC response

fredfortier committed:
DEV: Testing getActiveWorkers with the latest Enigma contract

fredfortier committed:
DEV: Working on Principal node signature verification.

fredfortier committed:
DEV: Working on integration with network data

fredfortier committed:
DEV: Integrated the Principal node to the Jest test suite

fredfortier committed:
FIX: Removed the public key from the `get_state_keys` response and included it in the non-encrypted part of the PrincipalMessage response struct

fredfortier committed:
DEV: Modified response log message

fredfortier committed:
DEV: Calling `setWorkerParams` in the epoch loop when a new epoch is found (more cleanup pending)

fredfortier committed:
DEV: Refactoring and cleanup around the `watch_blocks` business logic. Remembers the prev epoch block on shutdown.

fredfortier committed:
WIP: Working on selected workers cache

fredfortier committed:
DEV: Storing selected workers in with the epoch state

fredfortier committed:
DEV: Minor fixes

fredfortier committed:
FIX: Renamed `get_state_keys` to `getStateKeys` and updated the docs (#78)

fredfortier committed:
FIX: Renamed `get_state_keys` to `getStateKeys` and updated the docs (#78)

fredfortier committed:
FIX: Troubleshooting an issue with the report signature

fredfortier committed:
DEV: Worked on the registration workflow to avoid duplicate registrations between runs, validate the signer address, etc

fredfortier committed:
Merge remote-tracking branch 'origin/develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-core/enclave/Cargo.toml
#	enigma-crypto/src/error.rs
#	enigma-principal/app/Cargo.lock
#	enigma-principal/enclave/Cargo.toml
#	enigma-tools-t/src/common/errors_t.rs
#	enigma-tools-t/src/lib.rs
#	enigma-tools-u/Cargo.toml

fredfortier committed:
WIP: Merged the latest develop. Synced the boilerplace esgx code with enigma-core to fix an issue with the signing address data format. Made the register business logic more robust. Started incorporating logging. Incorporated the new PrincipalMessage format without the associated business logic.

fredfortier committed:
FIX: Fixed enigma-core code that I broke when merging

elichai committed:
MERGE: Cherry picked commit 0x5420f6Aec6878047C6FbBE69b035225aF4ea88be

fredfortier committed:
DEV: Using "submit with confirmations" and passing in the confirmations parameter

fredfortier committed:
WIP: Attempting to support the SW mode

fredfortier committed:
FIX: Added missing Cargo.lock files

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-core/enclave/Cargo.lock
#	enigma-principal/enclave/Cargo.lock

fredfortier committed:
WIP: Dependencies problems. Multiple Crates are broken with the 2018-09-30 compiler. This nightly seems to be missing multiple features that libraries expect in 1.3.31. Using the 2018-10-31 for now.

fredfortier committed:
DEV: First successful test of the simulation mode. Also, patched the Attestation Service until waiting for a more permanent fix.

fredfortier committed:
FIX: Adjusted the unicase library version

fredfortier committed:
WIP: Implementing message translator to add secret contract addresses to requests

fredfortier committed:
DEV: Implementing epoch address injection in getStateKeys requests.
WIP: Implementing epoch address injection in getStateKeys requests. Fixing bugs in the previous commit.
WIP: Implementing epoch address injection in getStateKeys requests
DEV: Refactoring away from the common modules.

fredfortier committed:
DEV: Added log macros

fredfortier committed:
FIX: Housekeeping to fix warnings
FIX: Housekeeping, cleaning-up imports
FIX: Housekeeping

elichai commented on eng-wasm/dispatch/src/lib.rs:
```diff
@@ -0,0 +1,256 @@
+#![recursion_limit="128"]
```
What's the deal with the `eng-wasm/dispatch`?  I think this is something from a merge got wrong.

fredfortier replied:
I hadn't noticed this. I will investigate.

fredfortier replied:
Fixed

elichai commented on enigma-core/app/enigma-core-app.iml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please add the `iml`  files to the `.gitignore`

fredfortier replied:
Fixed

elichai commented on enigma-principal/app/src/boot_network/keys_provider_http.rs:
```diff
@@ -0,0 +1,171 @@
+    #[logfn(DEBUG)]
+    fn get_state_keys_internal(epoch_provider: Arc<EpochProvider>, request: StateKeyRequest, eid: sgx_enclave_id_t) -> Result<Value, Error> {
+        println!("Got get_state_keys request: {:?}", request);
+        let reader = PrincipalMessageReader::new(request.data.clone().into())?;
```
the `Into` here contains an unwrap. this means that if you receive a non hex string in the JSON-RPC the principal node will crash.
I think the principal node should be a lot more robust, at least in the interface side.

fredfortier replied:
Good catch! Fixed in commit a2c322bc89442eaa0f2b175f400fcf53eb424988.

elichai commented on enigma-principal/app/src/epoch_u/epoch_provider.rs:
```diff
@@ -0,0 +1,267 @@
+pub struct EpochProvider {
+    pub contract: Arc<EnigmaContract>,
+    pub epoch_state: Arc<Mutex<Option<EpochState>>>,
+    pub eid: Arc<AtomicU64>,
```
I think the `eid` can be regular `u64` and doesn't need to be `AtomicU64` because there's no need to it to be atomic because it should never change.

fredfortier replied:
I wasn't able to pass a u64 across threads but will try again.

fredfortier replied:
Fixed a62f7c406935853558959cd57ceed2300154e3ce

elichai commented on enigma-principal/app/src/boot_network/keys_provider_http.rs:
```diff
@@ -0,0 +1,171 @@
+            Ok(body)
+        });
+        let server = ServerBuilder::new(io)
+            .cors(DomainsValidation::AllowOnly(vec![AccessControlAllowOrigin::Null]))
```
What is the need for CORS headers?

fredfortier replied:
The parameter was removing CORS header verification. I set it for troubleshooting as one point in time but you are correct that it's not needed. Removed in commit 5d2285cdfe8d20f05d7b9c88a11a173efc93ad4f.

elichai commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -63,117 +117,196 @@ impl PrincipalManager {
     fn get_contract_address(&self) -> Address { self.contract.address() }
 
-    fn get_quote(&self) -> Result<String, Error> { Ok(retry_quote(self.eid, &self.config.spid, 18)?) }
+    fn get_account_address(&self) -> Address { self.contract.account.clone() }
```
`Address` Implements `Copy` so no need to clone it.

fredfortier replied:
Interesting, I did not realize that implementing the `Copy` trait implicitly cloned the variable instead of moving. I thought that it still needed to be specified explicitly. My IDE does not seem to support this but the compiler is happy with it. I'll inspect throughout and create a single commit.

elichai commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -63,117 +117,196 @@ impl PrincipalManager {
-        signing_address = signing_address[2..].to_string();
-        Ok(signing_address)
+    fn get_block_number(&self) -> Result<U256, Error> {
+        let block_number = self.get_web3().eth().block_number().wait().unwrap();
```
Why do you unwrap if you return a result?

fredfortier replied:
Yes, I forgot to handle this. Fixed in commit 031fd8c363fa5688e64bd95558f29540aca50e14.

elichai commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -63,117 +117,196 @@ impl PrincipalManager {
+        let registration_params = self.report_manager.get_registration_params()?;
+        let report = registration_params.report.clone();
+        let signature = registration_params.signature.clone();
+        let receipt = self.contract.register(signing_address, report, signature, gas_limit, self.config.confirmations as usize)?;
```
```suggestion
        let receipt = self.contract.register(signing_address, registration_params.report, registration_params.signature, gas_limit, self.config.confirmations as usize)?;
```
You can pass the report and signature directly without cloning them in memory

fredfortier replied:
Fixed in commit: e7adf35481143153d4b637aa8f6ae93499172d2e.

elichai commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -63,117 +117,196 @@ impl PrincipalManager {
-        println!("signing address = {}", signer);
-        enigma_contract.register(&signer, &rlp_encoded, gas_limit)?;
+        // Start the WorkerParameterized Web3 log filter
+        let eid: Arc<AtomicU64> = Arc::new(AtomicU64::new(self.eid));
```
Again, the eid can stay u64 (or more clearly `sgx_enclave_id_t`)

fredfortier replied:
Fixed a62f7c406935853558959cd57ceed2300154e3ce

elichai commented on enigma-principal/app/src/boot_network/principal_manager.rs:
```diff
@@ -183,67 +316,62 @@ mod test {
     #[test]
-    //#[ignore]
+    #[ignore]
     fn test_full_principal_logic() {
```
Why is this test commented out?
I think it's a pretty important test, unless there's a different test that checks if random emission with simulated blocks

fredfortier replied:
Yes, I explained this in the notes included in the PR. That said, I agree that this test is useful. I just did not prioritize refactoring it since I have another way to test. I will re-enable it shortly.

fredfortier replied:
Fixed the unit test (minus contract deployment) 31888b32e95a4a7471e809442e465a89c9195c36

elichai commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,79 +1,101 @@
 pub fn start(eid: sgx_enclave_id_t) -> Result<(), Error> {
     let opt = cli::options::Opt::from_args();
-    let config = deploy_scripts::load_config(opt.deploy_config.as_str())?;
+    let _config = deploy_scripts::load_config(opt.deploy_config.as_str())?;
```
If you don't use this config why not remove it?

fredfortier replied:
Fixed in 68afcef246bbe4978d4ab94501853b45ead7fd67.

elichai commented on enigma-principal/app/src/cli/app.rs:
```diff
@@ -1,79 +1,101 @@
+    } else if opt.sign_address {
+        let mut path = storage_dir();
+        path.join(ENCLAVE_DIR);
+        path.push("principal-sign-addr.txt");
```
Why do you save the signing address to a file?

fredfortier replied:
This is important for integration testing. The contract deployment script reads the address from here.

elichai commented on enigma-tools-t/src/document_storage_t.rs:
```diff
@@ -0,0 +1,157 @@
+use sgx_tseal::SgxSealedData;
+#[cfg(not(target_env = "sgx"))]
```
You can remove this.
It's something old that we don't use.

fredfortier replied:
Removed in commit b9f50bc39689924bc12c3f0e14ee5792b72c861d.

elichai commented on enigma-tools-t/src/document_storage_t.rs:
```diff
@@ -0,0 +1,157 @@
+    pub fn unseal(sealed_log_in: &mut [u8]) -> Result<Option<Self>, EnclaveError> {
+        let sealed_log_size: usize = SEAL_LOG_SIZE;
+        let sealed_log = sealed_log_in.as_mut_ptr();
+        let sealed_data = match from_sealed_log::<Self>(sealed_log, sealed_log_size as u32) {
```
I think this should get a deeper look to see if we can use the "safer" ` unseal_data()`

fredfortier replied:
Most this file comes from the original `storage_t`. I haven't look at optimizations beyond variable length and error handling. I will do some research on this safer `unseal_data`.

elichai replied:
and it works? I'm actually pretty surprised, because the old `storage_t` was written before we really understood a lot and it's pretty hacky, so I'm surprised it works for bigger data

fredfortier replied:
Yes, it works for variable size of byte arrays. I'm not trying to store large documents at the moment. Do you have an example of a more mature implementation of storage which I can review?

elichai commented on enigma-tools-t/src/document_storage_t.rs:
```diff
@@ -0,0 +1,157 @@
+        )?;
+        let sealed_log = sealed_log_out.as_mut_ptr();
+        let sealed_log_size: usize = SEAL_LOG_SIZE;
+        to_sealed_log(&sealed_data, sealed_log, sealed_log_size as u32);
```
Is that really needed? can we just use the Self this returns?
I'll prefer if we didn't use the unsafe functions here.

fredfortier replied:
This is just an adaptation of `storage_t` which I assumed was current. I need to better understand the storage service to answer this. Is there a better example than `storage_t` in our code? If not, I research in the Baidu SDK.

elichai commented on enigma-tools-t/src/document_storage_t.rs:
```diff
@@ -0,0 +1,157 @@
+        // generate mock data
+        let mut doc: SealedDocumentStorage<[u8; 32]> = SealedDocumentStorage {
+            version: 0x1234,
+            data: [0; 32],
```
```suggestion
            data: [b'i'; 32],
```

elichai replied:
Then you can remove the for loop

fredfortier replied:
Obviously :smile: Commit 1bcbdd7da8841587d357af9b77be98dea69b83eb.

elichai commented on enigma-tools-t/src/document_storage_t.rs:
```diff
@@ -0,0 +1,157 @@
+
+use common::errors_t::EnclaveError;
+
+pub const SEAL_LOG_SIZE: usize = 2048;
```
Why is the whole 2048?
What if you sealed a `[u8; 8096]`?

fredfortier replied:
That's just the default value that was already. This should be enough for the data that I'm storing but I'll research this to see if can be more dynamic.

elichai commented on enigma-tools-t/Cargo.toml:
```diff
@@ -1,3 +1,5 @@
+cargo-features = ["rename-dependency"]
```
Why do you need this? I don't see any dependency that you're renaming

fredfortier replied:
Forgot to remove the feature after experimenting with it. Fixed in commit bed425e47186a9933bc4329680ffa9aeb31ad2cb.

elichai commented on enigma-principal/enclave/src/lib.rs:
```diff
@@ -59,50 +74,63 @@ fn get_sealed_keys_wrapper() -> asymmetric::KeyPair {
+                                                  sig_out: &mut [u8; 65]) -> EnclaveReturn {
+    let msg_bytes = slice::from_raw_parts(msg, msg_len).to_vec();
+    let addrs_bytes = slice::from_raw_parts(addrs, addrs_len).to_vec();
+    let response = match ecall_get_enc_state_keys_internal(msg_bytes, addrs_bytes, sig.clone(), sig_out) {
```
why clone the signature?

fredfortier replied:
Fixed in commit e7adf35481143153d4b637aa8f6ae93499172d2e.

elichai commented on enigma-principal/app/src/esgx/keys_keeper_u.rs:
```diff
@@ -0,0 +1,118 @@
+pub fn get_enc_state_keys(eid: sgx_enclave_id_t, request: StateKeyRequest, epoch_addrs: Option<Vec<H256>>) -> Result<StateKeyResponse, Error> {
+    let mut retval: EnclaveReturn = EnclaveReturn::Success;
+    let mut sig_out: [u8; 65] = [0; 65];
+    let mut enc_response = vec![0u8; MAX_ENC_RESPONSE_LEN];
```
What if it's higher than 100_000?
you should use: https://github.com/enigmampc/enigma-core-internal/blob/develop/enigma-core/enclave/src/ocalls_t.rs#L26
And then you unbox the memory, see this for reference: 
https://github.com/enigmampc/enigma-core-internal/blob/develop/enigma-core/app/src/km_u.rs#L65

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,242 @@
+            }
+            None => None,
+        };
+        if key.is_none() {
```
you can put in `Some(key) => results.push(key.clone())` and in None => everything you did in `if key.is_none()`

fredfortier replied:
I implemented it like this for a specific reason, putting the `key.is_none()` logic in the match fn gives an immutable (`guard get`) and mutable (`guard.insert`) borrow conflict.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,242 @@
+fn new_state_keys(guard: &mut SgxMutexGuard<HashMap<ContractAddress, StateKey, RandomState>>, sc_addrs: &Vec<ContractAddress>) -> Result<Vec<StateKey>, EnclaveError> {
+    let mut results: Vec<StateKey> = Vec::new();
+    for addr in sc_addrs {
+        let mut rand_seed: [u8; 1072] = [0; 1072];
```
Why do you generate a whole MegaByte of random just to take the first 32 bytes?

fredfortier replied:
This is how I've need in done in the existing code. I'll research further to better understand the `rand` function.

fredfortier replied:
Thanks for catching this. Fixed in commit 1df39b26a9e42561bd03defb2c5136408846a391.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,242 @@
+/// Get encrypted state keys
+pub(crate) fn ecall_get_enc_state_keys_internal(msg_bytes: Vec<u8>, addrs_bytes: Vec<u8>, sig: [u8; 65], sig_out: &mut [u8; 65]) -> Result<Vec<u8>, EnclaveError> {
+    let msg = PrincipalMessage::from_message(&msg_bytes)?;
+    let sc_addrs: Option<Vec<ContractAddress>> = match msg.data.clone() {
```
Why an option of vec and not just a vec and then you could return an empty one?

fredfortier replied:
You're right, the option is unnecessary and the code looks cleaner without. Commit e6592f92c553e998e5dbb454861702eb027b2d61.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,242 @@
+    Ok(results)
+}
+
+fn build_get_state_keys_response(_sc_addrs: Option<Vec<ContractAddress>>) -> Result<Vec<(ContractAddress, StateKey)>, EnclaveError> {
```
you can receive here a vec and just check if it's empty. no need for the Option.
this will remove an `unwrap` and make the code look nicer.

fredfortier replied:
Refactored in commit e6592f92c553e998e5dbb454861702eb027b2d61.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,242 @@
+    let response_data = build_get_state_keys_response(sc_addrs)?;
+
+    // Generate the encryption key material
+    let mut rand_num: [u8; 1072] = [0; 1072];
```
Again, a whole MB for just 44 bytes?

fredfortier replied:
Fixed in commit 1df39b26a9e42561bd03defb2c5136408846a391.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/keeper_types_t.rs:
```diff
@@ -0,0 +1,108 @@
+impl Decodable for InputWorkerParams {
+    fn decode(rlp: &UntrustedRlp) -> Result<Self, DecoderError> {
+        Ok(Self {
+            block_number: U256::from_bigint(rlp.val_at(0)?),
```
What's up with these bigints?
I can't see them in the docs of etccommon-rlp, where do they come from? do we really need to have bigint as a dependency? https://docs.rs/etcommon-rlp/0.2.4/rlp/trait.Decodable.html

fredfortier replied:
It's required to use this particular ectcommon RLP Crate. These crates use `bigint` instead of `ethereum-types`. I suppose that I could use a different encoding method here since RLP is no longer imperative.

elichai replied:
hmm ok, I still haven't looked into where is this used but I also don't like the endianess assumeption here.
what if `ethereum-types::U256` and `etcommon-bigint::U256` don't use the same endianess?

elichai commented on enigma-principal/enclave/src/epoch_keeper_t/mod.rs:
```diff
@@ -0,0 +1,176 @@
+    let nonce_bytes: EpochNonce = nonce.into();
+    nonce_out.copy_from_slice(&nonce_bytes[..]);
+
+    // TODO: Check if needs to check the random is within the curve.
```
curve?

fredfortier replied:
This comes from the original code. I believe that Isan wrote this comment. I don't understand it  either but did not want to lose track of it.

fredfortier replied:
Removed the TODO in 3d1f5d0f650db92fb9f3f0ff3125074c497d4cc6.

elichai commented on enigma-principal/enclave/src/epoch_keeper_t/mod.rs:
```diff
@@ -0,0 +1,176 @@
+    let msg = epoch.raw_encode()?;
+    let hash = msg.keccak256();
+    println!("Signing msg hash {} with signer address {}", hash.to_hex(), SIGNING_KEY.get_pubkey().address_string());
+    let sig = SIGNING_KEY.sign(hash.as_ref())?;
```
You know that the sign message already hash the data. right?

fredfortier replied:
Ah! Thanks for reminding me. This explain the issue. I'm not sure how I got it mixed-up.

fredfortier replied:
Removed the redundant hashing in commit 3d1f5d0f650db92fb9f3f0ff3125074c497d4cc6.

elichai commented on enigma-principal/enclave/src/epoch_keeper_t/mod.rs:
```diff
@@ -0,0 +1,176 @@
+/// Creates new epoch both in the cache and as sealed documents
+fn new_epoch(guard: &mut SgxMutexGuard<HashMap<Uint, Epoch, RandomState>>, worker_params: &InputWorkerParams, nonce: &Uint, seed: &Uint) -> Result<Epoch, EnclaveError> {
+    let mut marker_doc: SealedDocumentStorage<EpochNonce> = SealedDocumentStorage {
+        version: 0x1234, //TODO: what's this?
```
Right. what is this? :)

fredfortier replied:
I have no idea. This comes from the original `storage_t`. I tried not to break things that I did not understand. Isan might know.

elichai replied:
I think this is a feature that @Isan-Rivkin thought might be useful for the signing keys, I don't think there's a use case here and even in the signing keys we didn't end up using this.

elichai commented on enigma-principal/enclave/src/epoch_keeper_t/mod.rs:
```diff
@@ -0,0 +1,176 @@
+}
+
+fn get_max_nonce(guard: &SgxMutexGuard<HashMap<Uint, Epoch, RandomState>>) -> Uint {
+    guard.keys().max().unwrap().clone()
```
Can the HashMap be empty?
If so maybe do `unwrap_or_deafult()` to return `0` in the case it's empty?

fredfortier replied:
No, it can't, there is a guard clause before its usage. Since this fn was only used once, I refactored it to put the logic directly after the guard clause. I can see how this wasn't very readable. Commit dd39e025a5b34eecd022a5071f6588798cdc49cd.

elichai commented on enigma-principal/app/src/logging.rs:
```diff
@@ -0,0 +1,38 @@
+use std::{fs::{self, File}, path::Path};
```
maybe this file can be moved to `enigma-tools-u`

fredfortier replied:
Refactored in commit f548eafbf818bd9c44486a3f4b0d673971e97afc. There is more code copy/pasted especially in `enigma-principal/app/src/esgx` module. I tried to not to question functioning code and limit impact on shared modules, but it's probably worth refactoring this area.

elichai commented on enigma-principal/app/src/keys_u/km_reader.rs:
```diff
@@ -0,0 +1,107 @@
+}
+
+#[derive(Debug, PartialEq, Clone, Deserialize)]
+pub struct PrincipalMessage {
```
Why are you deserializing this outside of the enclave?
Why not pass it directly to the enclave and deserialize it && verify the signature there?

fredfortier replied:
Here is the logic:
1. The list of contract => selected worker is stored on the untrusted side
2. Here, I read the message to verify if the list of addresses is None. If so, I find the contract addresses for this worker in the epoch and pass to the enclave along with the message.
3. If the message does contain a list of addresses, I just pass the message to the enclave.
4. In the enclave, I deserialize the message again and read the addresses either from the message or, in None, from the list passed by the untrusted. I also verify the signature and run the worker selection algorithm to confirm that the addresses are valid.

This approach achieves two things:
 1. Allow messages to pass-through the the enclave so that the untrusted side cannot censor the addresses that it may contain.
2. Keeps the list of secret contracts on the untrusted side. It would be hard to prove this list to the enclave. At most, the untrusted side could censor secret contract addresses selected for the worker. However, the worker could simply follow-up with a message requesting the missing addresses which the untrusted side cannot modify.

elichai commented on enigma-principal/app/src/epoch_u/epoch_types.rs:
```diff
@@ -0,0 +1,140 @@
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(3);
+        s.append(&self.block_number.bigint());
+        s.append_list(&self.workers.iter().map(|a| a.bigint()).collect::<Vec<bigint::H160>>());
```
Again, I can't seem to find where the bigint comes from.
https://docs.rs/etcommon-rlp/0.2.4/rlp/trait.Encodable.html

fredfortier replied:
Sam answer as above. It's the types primitive for etccommon modules.

elichai reviewed (changes requested):
Ok, this is my first pass of the code.
It's really really hard to review such a big PR. 
next time please try to split into little PRs,
this will help you with the need to always merge from develop and will help me in reviewing.

elichai commented on enigma-tools-u/src/web3_utils/w3utils.rs:
```diff
@@ -74,6 +74,11 @@ pub fn connect(url: &str) -> Result<(web3::transports::EventLoopHandle, Web3<Htt
 }
 
+pub fn connect_batch(http: Http) -> Web3<Batch<Http>> {
+    let w3 = web3::Web3::new(web3::transports::Batch::new(http));
```
```suggestion
    web3::Web3::new(web3::transports::Batch::new(http))
```
You can return this directly

fredfortier replied:
You're right. I'm not using the batch transport anymore. I will clean-up.

fredfortier replied:
Removed `connect_batch` in commit b9f50bc39689924bc12c3f0e14ee5792b72c861d.

elichai commented on enigma-principal/app/src/esgx/keys_keeper_u.rs:
```diff
@@ -0,0 +1,118 @@
+    if retval != EnclaveReturn::Success || status != sgx_status_t::SGX_SUCCESS {
+        return Err(EnclaveFailError { err: retval, status }.into());
+    }
+    let enc_response_out: Vec<u8> = enc_response_slice[0..enc_response_len_out].iter().cloned().collect();
```
```suggestion
    let enc_response_out: Vec<u8> = enc_response_slice[0..enc_response_len_out].to_vec()
```
This should be faster and is more readable

elichai commented on enigma-principal/app/src/epoch_u/epoch_provider.rs:
```diff
@@ -0,0 +1,267 @@
+        let epoch_state = &mut set_worker_params(self.eid.load(Ordering::SeqCst), worker_params.clone())?;
+        println!("Waiting for setWorkerParams({:?}, {:?}, {:?})", block_number, epoch_state.seed, epoch_state.sig);
+        // TODO: Consider a retry mechanism, either store the EpochSeed or add a getter ecall
+        let receipt = self.contract.set_workers_params(block_number, epoch_state.seed.clone(), epoch_state.sig.clone(), gas_limit, confirmations)?;
```
No need to clone the seed, it's a `Copy` type

fredfortier replied:
Adjusted in commit 762ae1c73e8e00ab8c54fcc6c4e9c5eb788d1d32.

elichai commented on enigma-principal/app/src/keys_u/keeper_types_u.rs:
```diff
@@ -0,0 +1,118 @@
+            let mut rand_val: U256 = U256::from(hash) % balance_sum;
+            println!("The initial random value: {:?}", rand_val);
+            let mut selected_worker = self.workers[self.workers.len() - 1].clone();
+            for i in 0..self.workers.len() {
```
why not just `for worker in self.workers` ?

fredfortier replied:
Because I need the index in the line below.

elichai replied:
Ok, FYI best practice is `enumerate` https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.enumerate

elichai committed:
Update enigma-principal/app/src/esgx/keys_keeper_u.rs

Co-Authored-By: fredfortier <fredfortier@gmail.com>
FIX: Housekeeping and improving docstrings
FIX: Bug fixes

elichai committed:
BLD: Fixed the prinicpal Makefile

fredfortier committed:
FIX: Removing extraneous `eng_wasm` dispatch module

fredfortier committed:
FIX: Removing extraneous iml files

fredfortier committed:
FIX: Improved error handling of the get_state_keys interface

elichai commented on dockerfile/run.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
I'm not sure if that's the right place for this. e.g. I just had this in my `.bash_aliases`/`.bashrc`:
```
alias enigma-docker='docker run --security-opt seccomp:unconfined --cap-add=SYS_PTRACE --net="host" -v /home/elichai2/CLionProjects/enigma-core:/home/elichai2/enigma-core -v /home/elichai2/.cargo-enigma-docker:/home/elichai2/.cargo -ti --device /dev/isgx enigma_dev:latest'
```
So I don't think that your exact preferences should be in the git

elichai commented on .gitignore:
```diff
@@ -85,3 +86,9 @@ fabric.properties
+**/*.rs.bk
+
+/target
+**/*.rs.bk
```
You wrote this twice.

fredfortier replied:
Fixed 90deb885612469ce15129c979a16cabe78c09549

elichai commented on enigma-principal/enclave/src/lib.rs:
```diff
@@ -59,50 +74,63 @@ fn get_sealed_keys_wrapper() -> asymmetric::KeyPair {
 
 #[no_mangle]
 pub extern "C" fn ecall_get_signing_address(pubkey: &mut [u8; 42]) {
+    println!("The priv key for worker 8: {:?}", SIGNING_KEY.get_privkey().to_vec().to_hex());
```
I'm gessing you meant to remove this.

fredfortier replied:
Fixed c8ecac858baab8e1445c524e6103b2c6df3d3f76

fredfortier committed:
FIX: Fixed PR requests - AtomicU64, TryInto, setWorkerParams, etc.
FIX: Replaced redundant AtomicU64
FIX: Removed duplicate entries
FIX: Remove unnecessary log file
FIX: Removing the priv key print
FIX: Fixed the TryInto implementation
FIX: Removed unnecessary code in the document storage unit test
FIX: Housekeeping to improve the readibility of functions in the keys_keeper module
FIX: Fixed the setWorkerParams message signature
FIX: Housekeeping in the epoch keeper
FIX: Fix rand byte array length and other fixes
DEV: Moved the logging module to enigma_tools_u
FIX: Removed redundant clone of a variable implementing Copy
FIX: Removed the `connect_batch` fn which is no longer used
FIX: Remove unnecessary config
FIX: Removed unnecessary feature
FIX: Housekeeping in the app file
FIX: Removed redundant clone calls
FIX: Improved error handling when fetching the current block number
FIX: Removed explicit CORS domain validation config

moriaab commented on enigma-principal/app/src/boot_network/keys_provider_http.rs:
```diff
@@ -0,0 +1,174 @@
+        let value = &self;
+        let bytes = value.0.from_hex()?;
+        let mut slice: [u8; 65] = [0; 65];
+        slice.copy_from_slice(&bytes[..]);
```
The length of source must be the same as self in copy_from_slice, the function will panic otherwise. String can be any length. You should check it somehow or do it safer.

moriaab commented on enigma-principal/app/src/main.rs:
```diff
@@ -1,34 +1,61 @@
+mod keys_u;
 
 #[allow(unused_variables, unused_mut)]
 fn main() {
```
@elichai, @fredfortier
This is a general comment. We have duplicated code in functions `main` in enigma-core/app/src/main.rs and enigma-principal/app/src/main.rs: from the first line to getting eid. I think it is better to extract it to untrusted common utils.

elichai commented on enigma-principal/app/src/esgx/equote.rs:
```diff
@@ -1,52 +1,57 @@
 
 extern "C" {
-    fn ecall_get_signing_address(eid: sgx_enclave_id_t, pubkey: &mut [u8; 42]) -> sgx_status_t;
+    fn ecall_get_signing_address(eid: sgx_enclave_id_t, pubkey: &mut [u8; 20]) -> sgx_status_t;
```
You changed the signature here but didn't change it in the EDL or in the trusted side.
This means the edgerator just memcpy's another 22 bytes that just exist after this array. I'm suprised you didn't have any segmentation faults.

fredfortier replied:
The trusted side looks fine but I fixed the mismatching EDL f58c88976c52619e382b724ddb31587d204d5dad

elichai replied:
That's because you fixed it here :) https://github.com/enigmampc/enigma-core-internal/pull/89/commits/45a294ae75c62b62b91020fd053521727a4f7545

fredfortier committed:
FIX: Fixed the epoch block number stored in the state

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,241 @@
+    // Generate the iv from the first 12 bytes of a new random number
+    let mut iv: [u8; 12] = [0; 12];
+    iv.clone_from_slice(&rand_num[32..44]);
+    let response = response_msg.encrypt_with_nonce(&derived_key, Some(iv))?;
```
`encrypt_with_nonce` shouldn't be called directly, it's for internal use and tests.
I'll make a PR that adds a warning there.

fredfortier replied:
I don't see any abstraction in the `km_primitive` module.

elichai replied:
I wrote that trait in a way that you need to implement `encrypt_with_nonce` and you get `encrypt` for free https://github.com/enigmampc/enigma-core-internal/blob/develop/enigma-crypto/src/lib.rs#L37

elichai commented on enigma-principal/enclave/src/lib.rs:
```diff
@@ -62,47 +77,59 @@ pub extern "C" fn ecall_get_signing_address(pubkey: &mut [u8; 42]) {
+                                                  enc_response_out: *mut u8, enc_response_len: &mut usize,
+                                                  sig_out: &mut [u8; 65]) -> EnclaveReturn {
+    let msg_bytes = slice::from_raw_parts(msg, msg_len).to_vec();
+    let addrs_bytes = slice::from_raw_parts(addrs, addrs_len).to_vec();
```
No need to convert these to vectors, you can pass them as slices directly.

fredfortier replied:
My compiler wants me to. I'm not sure why it's using a `Vec` implementation. I believe that I'm importing it correctly.
```
error[E0308]: mismatched types
   --> src/lib.rs:103:60
    |
103 |     let response = match ecall_get_enc_state_keys_internal(*msg_bytes, *addrs_bytes, *sig, sig_out) {
    |                                                            ^^^^^^^^^^
    |                                                            |
    |                                                            expected struct `serde::export::Vec`, found slice
    |                                                            help: try using a conversion method: `*msg_bytes.to_vec()`
    |
    = note: expected type `serde::export::Vec<u8>`
               found type `[u8]`
```

elichai replied:
Change `ecall_get_enc_state_keys_internal` to accept slices

elichai commented on enigma-principal/app/src/esgx/keys_keeper_u.rs:
```diff
@@ -0,0 +1,120 @@
+            &mut retval,
+            msg_bytes.as_ptr() as *const u8,
+            msg_bytes.len(),
+            addrs_bytes.as_ptr() as *const u8,
```
We should always use `as_c_ptr()`(https://github.com/enigmampc/enigma-core-internal/blob/develop/enigma-types/src/traits.rs#L3) otherwise if the slice/vec is empty it will return a ptr 0x0, which isn't a defined behavior in C, so this causes SegFaults.

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,241 @@
+    rsgx_read_rand(&mut rand_num)?;
+    let mut privkey_slice = [0u8; 32];
+    privkey_slice.copy_from_slice(&rand_num[..32]);
+    let key_pair = KeyPair::from_slice(&privkey_slice)?;
```
Please use `KeyPair::new()` as the generated private key might not be valid for the curve. it should be checked (which `KeyPair::new()` does)

fredfortier replied:
Fixed 951089c26472ff38de1975e13f5c7bb492e6db27

fredfortier committed:
FIX: Added a guard condition to handle mismatching request attribute size

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,241 @@
+    let id = msg.get_id();
+    let pubkey = key_pair.get_pubkey();
+    let response_msg = PrincipalMessage::new_id(response_msg_data, id, pubkey);
+    if !response_msg.is_response() {
```
Why do you check if it's not a response right after you made it as a response?

fredfortier replied:
I was trying to account for the possibility that the message is not initialized correctly. After looking more closely at what `new_id` does, I agree that it is redundant. Fixed 951089c26472ff38de1975e13f5c7bb492e6db27

elichai commented on enigma-principal/enclave/src/keys_keeper_t/mod.rs:
```diff
@@ -0,0 +1,241 @@
+    println!("The partially encrypted response: {:?}", response_bytes.to_hex());
+    // Signing the encrypted response
+    // This is important because the response might be delivered by an intermediary
+    let hash = response_bytes.clone().keccak256();
```
You're hashing before the sign function which already hashes for you.

fredfortier replied:
Fixed 951089c26472ff38de1975e13f5c7bb492e6db27

fredfortier committed:
DEV: Adjust the unit tests and contracts to use the signing in the worker selection algo

fredfortier committed:
FIX: Removed the address log message

fredfortier committed:
FIX: Fixed the full logic unit test

fredfortier committed:
FIX: Fixed a mismatching variable in the EDL

fredfortier committed:
FIX: Fixed some enigma-crypto calls in the key keeper

elichai committed:
DEV: Created a shared ocalls file in enigma-tools-u

elichai committed:
DEV: Created a shared ocalls_t in enigma-tools-t

elichai committed:
DEV: Edited the `ecall_get_enc_state_keys` function

elichai committed:
MAINT: Refactored epoch_keeper_t and keys_keeper_t

elichai committed:
DEV: Refactored the principal CLI and the epoch_u files

elichai committed:
MAINT: Refactoring the enigma-tools-u and other tiny stuff

elichai committed:
MAINT: Refactored the worker selection algorithm

elichai committed:
DEV: Removed `ocall_get_home` and `ocall_save_to_memory` from enigma-core and used the ones from tools

fredfortier committed:
FIX: Included addresses in the Principal message

elichai committed:
DEV: Replace H256 with ContractAddress and change remove the abi encoding for the edl

elichai committed:
DEV: Replace H256 with ContractAddress and change remove the abi encoding for the edl

fredfortier committed:
DEV: Added a CLI command for getStateKeys

fredfortier committed:
DEV: Added a CLI option to reset the Epoch state

fredfortier committed:
DEV: Testing getStateKeys from Enigma.spec.js. All tests passing.

elichai committed:
DEV: Created a tools crate to be shared between trusted and untrusted sides

elichai committed:
DEV: Moved keeper_types to a mutual tools crate

elichai committed:
DEV: Moved km_primitives into enigma-tools-m

elichai committed:
DEV: Updated the location of km_primitives and implemented Serialize/Deserialize to it.

elichai committed:
DEV: Moved EthereumAddress to enigma-tools-m

elichai committed:
DEV: Updated the enigma-core parts to enigma-tools-m

elichai committed:
DEV: Updated the serde-impls and fixed some tests

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/Cargo.lock
#	enigma-principal/app/src/boot_network/config.json
#	enigma-principal/app/src/boot_network/principal_manager.rs
#	enigma-principal/app/src/esgx/equote.rs
#	enigma-tools-t/src/common/errors_t.rs
#	enigma-tools-t/src/storage_t.rs
#	enigma-tools-u/Cargo.toml
#	enigma-tools-u/src/attestation_service/service.rs

fredfortier committed:
MERGE: Fixes post develop merge

fredfortier committed:
Merge branch 'km-changes' of github.com:enigmampc/enigma-core-internal into km-changes

fredfortier committed:
FIX: Disabling the integration tests for the time being

fredfortier committed:
FIX: Changes to lock files

fredfortier committed:
Merge branch 'develop' into principal-keymgmt

# Conflicts:
#	enigma-core/app/src/main.rs

fredfortier committed:
FIX: Post develop merge fixes

elichai committed:
Revert "FIX: Post develop merge fixes"

This reverts commit 1cd35e3a2175d9024416fcf25f9f8eec9077eb89.

elichai committed:
BLD: Added enigma-tools-m to the tests and updated Cargo.locks

fredfortier committed:
Merge branch 'enigma-tools-m' into principal-keymgmt

elichai committed:
MAINT: Rustfmt on enigma-principal/app and removed unnecessary dependencies

elichai committed:
MAINT: Rustfmt enigma-principal/enclave and removed unnecessary dependencies

elichai committed:
MAINT: Updated rustfmt.toml and Cargo.lock

fredfortier committed:
Merge branch 'rustfmt' into principal-keymgmt
